### PR TITLE
Feature: Semantic search

### PR DESCRIPTION
**Pull request by akshay9**
Used clip embedding to implement Semantic search

### CODE REVIEW

File: api/models/photo.py [hunk 1]
```diff
@@ -66,6 +68,8 @@ class Photo(models.Model):
 
     public = models.BooleanField(default=False, db_index=True)
     encoding = models.TextField(default=None, null=True)
+    clip_embeddings = ArrayField(models.FloatField(blank=True, null=True), size=512)
+    clip_embeddings_magnitutde = models.FloatField(blank=True, null=True)
```
Comment (derneuere) [inline api/models/photo.py]: Typo: Should be  clip_embeddings_magnitude

File: api/models/photo.py [hunk 2]
```diff
@@ -101,6 +105,22 @@ def _generate_captions_im2txt(self,commit=True):
                 'could not generate im2txt captions for image %s' % image_path)
             return False
 
+    def _generate_clip_embeddings(self, commit):
+        image_path = self.image_paths[0]
+
+        try:
+            img_emb, magnitutde = semantic_search_instance.calculate_clip_embeddings(image_path)
```
Comment (derneuere) [inline api/models/photo.py]: Typo: Should be magnitude

**Comment by derneuere** [comment]
Thank you for your contribution. Very cool feature! It works, but I found a couple of user experience issues.

1. Album Things uses the search to find the associated photos. Not sure why, but this pr affects the current implementation. This should be changed to just show the images of the Album Thing.
2. The search times out when scanning, probably because CPU is already utilized heavily. We could use the other search function while it's scanning.
3. I think the similarity_factor should be higher, to get less false positives. 

Because this is a very performance hungry feature, it would be great if you could enable / disable it in the settings via the UI. Some people want to run LibrePhotos on a Raspberry Pi in the future 😅 The similarity factor could also be a setting.

File: api/models/photo.py (hunk 1 above)
Comment (akshay9) [reply api/models/photo.py]: Fixed in commit be1dab2

File: api/models/photo.py (hunk 2 above)
Comment (akshay9) [reply api/models/photo.py]: Fixed in be1dab2

**Comment by akshay9** [comment]
Thank you @derneuere for you feedback. I have made some additional enhancements so that it's easier to use Semantic Search  on low powered devices or opt-out based on users' preference.
- Added a settings in the Settings area for user to enable/disable Semantic Search. Default setting is Disabled since semantic search is for use in high end devices only (at least >4 cores 8GB Ram)
- Initial clip embedding calculations is done after the setting is first enabled. Leading to faster migrations on upgrade.
- Clip Embedding is now calculated in Batches of 64 Rows, gaining performance boost of >100% compared to calculating clip embeddings for just 1 image per batch

File: api/image_similarity.py [hunk 3]
```diff
@@ -5,19 +5,40 @@
...
     if type(user) == int:
         user_id = user
     else:
         user_id = user.id
     
-    if(photo.encoding == None):
-        photo._im2vec()
-    if(photo.encoding == None):
+    if(photo.clip_embeddings == None):
+        photo._generate_clip_embeddings()
```
Comment (derneuere) [inline api/image_similarity.py:36]: This line leads to an exception, because the parameter commit is missing.

File: api/models/photo.py [hunk 4]
```diff
@@ -101,6 +103,22 @@ def _generate_captions_im2txt(self,commit=True):
                 'could not generate im2txt captions for image %s' % image_path)
             return False
 
+    def _generate_clip_embeddings(self, commit):
+        image_path = self.image_paths[0]
```
Comment (derneuere) [inline api/models/photo.py]: You should use thumbnail_big instead. thumbnail_big is always a JPG, and it is uncommon that a thumbnail is missing. The original picture could have been moved or could be in a file format that is not well understood by semantic_search_instance.calculate_clip_embeddings like a raw image or a video file.

**Comment by derneuere** [comment]
I found two edge cases which don't work correctly. Otherwise, it works really well and as one would expect. Awesome work 👍

File: api/image_similarity.py (hunk 3 above)
Comment (akshay9) [reply api/image_similarity.py:36]: Added a default value to kwarg

File: api/models/photo.py (hunk 4 above)
Comment (akshay9) [reply api/models/photo.py]: Oh! I wasn't aware of that, thanks a lot for this minute observation, I have changed it to thumbnail_big.path

**Comment by akshay9** [comment]
Thanks @derneuere , I have fixed those two edge cases

**Pull request merged**